consensus/src/lib.rs

```diff
@@ -116,38 +130,112 @@ impl<B, E, Block, RA> LocalClient for Client<B, E, Block, RA> where
 	}
 }
 
-fn parachain_key(para_id: ParaId) -> substrate_primitives::storage::StorageKey {
-	const PREFIX: &[u8] = &*b"Parachains Heads";
-	para_id.using_encoded(|s| {
-		let mut v = PREFIX.to_vec();
-		v.extend(s);
-		substrate_primitives::storage::StorageKey(v)
-	})
-}
-
 impl<B, E, RA> PolkadotClient for Arc<Client<B, E, PBlock, RA>> where
 	B: Backend<PBlock, Blake2Hasher> + Send + Sync + 'static,
 	E: CallExecutor<PBlock, Blake2Hasher> + Send + Sync + 'static,
-	RA: ProvideRuntimeApi + Send + Sync + 'static,
-	RA::Api: ParachainHost<PBlock>,
+	Client<B, E, PBlock, RA>: ProvideRuntimeApi + Send + Sync + 'static,
+	<Client<B, E, PBlock, RA> as ProvideRuntimeApi>::Api: ParachainHost<PBlock>,
 {
 	type Error = ClientError;
 
 	type Finalized = Box<dyn Stream<Item=Vec<u8>> + Send + Unpin>;
 
 	fn finalized_heads(&self, para_id: ParaId) -> ClientResult<Self::Finalized> {
 		let polkadot = self.clone();
-		let parachain_key = parachain_key(para_id);
 
 		let s = self.finality_notification_stream()
 			.filter_map(move |n|
 				future::ready(
-					polkadot.storage(&BlockId::hash(n.hash), &parachain_key)
-						.ok()
-						.and_then(|d| d.map(|d| d.0)),
+					polkadot.parachain_head_at(&BlockId::hash(n.hash), para_id).ok().and_then(|h| h),
 				),
 			);
 
 		Ok(Box::new(s))
 	}
+
+	fn parachain_head_at(
+		&self,
+		at: &BlockId<PBlock>,
+		para_id: ParaId,
+	) -> ClientResult<Option<Vec<u8>>> {
+		self.runtime_api().parachain_status(at, para_id).map(|s| s.map(|s| s.head_data.0))
```

Comment:
IMO querying storage is much more light-client friendly.
Alternatively we could scrape the block body.